torch/csrc/jit/codegen/cuda/runtime/block_welford_outer.cu

```diff
@@ -0,0 +1,265 @@
+    const DataType b_N_div_ab_N =
+        updated_N != 0 ? ((DataType)peer_N) / ((DataType)updated_N) : 0;
+
+#pragma unroll
+    for (int index_in_chunk = 0; index_in_chunk < chunk_size;
+         ++index_in_chunk) {
+      DataType pushed_avg = 0;
+      DataType pushed_var = 0;
+      DataType self_avg = 0;
+      DataType self_var = 0;
+      // Divergent branch. Not a big deal with independent scheduling?
+      if (lane_id & lane_mask) {
+        // Push first half
+        auto push_offset = index_in_chunk;
+        auto self_offset = index_in_chunk + chunk_size;
+        pushed_avg = inp_avg[push_offset];
+        pushed_var = inp_var[push_offset];
+        self_avg = inp_avg[self_offset];
+        self_var = inp_var[self_offset];
+      } else {
+        // Push second half
+        auto push_offset = index_in_chunk + chunk_size;
+        auto self_offset = index_in_chunk;
+        pushed_avg = inp_avg[push_offset];
+        pushed_var = inp_var[push_offset];
+        self_avg = inp_avg[self_offset];
+        self_var = inp_var[self_offset];
+      }
+      auto peer_avg = __shfl_xor_sync(0xffffffff, pushed_avg, lane_mask);
+      auto peer_var = __shfl_xor_sync(0xffffffff, pushed_var, lane_mask);
+
+      auto delta = peer_avg - self_avg;
+      self_avg += delta * b_N_div_ab_N;
+      self_var += peer_var + delta * delta * ((DataType)(inp_N)) * b_N_div_ab_N;
+
+      inp_avg[index_in_chunk] = self_avg;
+      inp_var[index_in_chunk] = self_var;
+    }
+    inp_N = updated_N;
+  }
+
+  // At this point, chunk_size is reduced to NumVals/wdimy as
+  // mentioned above. Each thread has warp-reduced chunk_size values
+  // in array inp. This chunk_size_post_reduction should be equal to
+  // chunk_size at this point.
+  constexpr int chunk_size_post_reduction = NumVals / wdimy;
+
+  // More specifically, the warp_tidy of each thread defines
+  // the chunk IDs held by the thread as follows:
+  //
+  // [warp_tidy * chunk_size_post_reduction, warp_tidy *
+  // chunk_size_post_reduction + chunk_size_post_reduction]
+  //
+  // Each thread uploads the chunk_size_post_reduction values one by
+  // one. Each chunk is spread by BDIMX * BDIMY values. The data
+  // layout of the shared memory is:
+  //
+  // [chunk_size, wid, warp_tidy, TIDx]
+  //
+  // The remaining reduction is done on the WID
+  // dimension. More specifically, we assign one warp per chunk (or
+  // a value of the group). The wdimy threads of the same threadId.x
+  // collectively reduce num_warps partial results, each of which is
+  // stored with stride 32. This means that there will be wdimy-way
+  // bank conflicts, so to avoid that, swizziling is also employed.
+#pragma unroll
+  for (int i = 0; i < chunk_size; ++i) {
+    // Accumulating smem offset from the innermost dimension
+    int smem_offset = 0;
+    // TIDx
+    smem_offset += threadIdx.x;
+    // Warp_TIDy with swizzle
+    smem_offset += ((warp_tidy + wid) % wdimy) * BDIMX;
```

Comment:
I don't understand `warp_tidy + wid` in this expression. That seems strange to me since `warp_tidy` is the y-index within the warp, and `wid` is the warp index. I guess I just wouldn't expect an addition here even with a swizzle as it seems not like a 1:1 mapping. What am I missing here?

Reply:
This can be considered as a 2D transpose pattern of a `num_warps` by `wdimy` tile. Each store operation to the shared memory buffer by a warp writes `wdimy` values for each non-reduction domain value, and each warp has `BDIMX` non-iteration domain values. This store is done by every warp, so there will be `num_warps` stored results. When reading the tile, the reduction is done along the `num_warps` dimension, not the `wdimy` dimension. Since a single warp is used for each group entry, the first `BDIMX` threads will read values of offset 
0 in the `wdimy` dimension, and the next `BDIMX` threads will read at offset `wdimy`, which means they fall in the same banks as those used by the first `BDIMX` threads, and so on.

This is effectively reducing each column of a row-major 2D tile of `num_warps` rows and `wdimy` columns, and `wid` and `warp_tidy` are the row and column indices, respectively. Thus, to avoid bank conflicts, the transpose swizzle is used as `((warp_tidy + wid) % wdimy)`